security/advancedtls/advancedtls.go

```diff
@@ -0,0 +1,377 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package advancedtls is a utility library containing functions to construct
+// credentials.TransportCredentials that can perform credential reloading and custom
+// server authorization.
+// All APIs in this package are EXPERIMENTAL.
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"syscall"
+	"time"
+
+	"google.golang.org/grpc/credentials"
+)
+
+type sysConn = syscall.Conn
+
+// CustomVerificationFunc is the function defined by users to perform custom server
+// authorization.
+type CustomVerificationFunc func(serverName string,
+	rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
+
+// RootCertificateOptions contains a field and a function for obtaining root trust certificates.
+// It is used by both ClientOptions and ServerOptions.
+type RootCertificateOptions struct {
+	// ---------------General Rules For Root Certificate Setting------------------------------------
+	// RootCACerts or GetRootCAs indicates the the certificates trusted by the client side. The rules
+	// for setting these two fields are:
+	// Either RootCACerts or GetRootCAs must be set; the other will be ignored
+	//----------------------------------------------------------------------------------------------
+	// If field RootCACerts is set, field GetRootCAs will be ignored. The client will use RootCACerts
+	// every time when verifying the peer certificates, without performing root certificate reloading.
+	RootCACerts *x509.CertPool
+	// If GetRootCAs is set and RootCACerts is nil, the client will invoke this function every time
+	// asked to check certificates sent from the server when a new connection is established.
+	// This is known as root CA certificate reloading.
+	GetRootCAs func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
+}
+
+// ClientOptions contains all the fields and functions needed to be filled by the client.
+type ClientOptions struct {
+	// ---------------General Rules For Certificate Setting-----------------------------------------
+	// Certificates or GetClientCertificate indicates the certificates sent from the client to the
+	// server to prove client's identities. The rules for setting these two fields are:
+	// If requiring mutual authentication on server side:
+	//     Either Certificates or GetClientCertificate must be set; the other will be ignored
+	// Otherwise:
+	//     Nothing needed(the two fields will be ignored)
+	//----------------------------------------------------------------------------------------------
```

Comment:
Similar: I think if a comment generally covers multiple fields then it should be with the struct and not any individual field.